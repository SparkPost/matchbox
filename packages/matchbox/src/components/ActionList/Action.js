import React from 'react';
import PropTypes from 'prop-types';
import { CheckBox } from '@sparkpost/matchbox-icons';
import { deprecate } from '../../helpers/propTypes';
import { Box } from '../Box';
import { HelpText } from '../HelpText';
import { StyledLink } from './styles';

const Action = React.forwardRef(function Action(props, userRef) {
  const { content, children, disabled, helpText, is = 'link', selected, ...action } = props;

  const linkContent = React.useMemo(() => {
    return (
      <Box as="span" alignItems="flex-start" display="flex">
        <Box as="span" flex="1" fontSize="400" lineHeight="300">
          {content || children}
        </Box>
        {selected && (
          <Box as="span" color="blue.700">
            <CheckBox size={20} />
          </Box>
        )}
        {helpText && <HelpText>{helpText}</HelpText>}
      </Box>
    );
  }, [content, selected, helpText]);

  return (
    <StyledLink
      as={is === 'button' ? 'button' : null}
      type={is === 'button' ? 'button' : null}
      disabled={disabled}
      isType={is}
      ref={userRef}
      role="menuitem"
      tabIndex="-1"
      {...action}
    >
      {linkContent}
    </StyledLink>
  );
});

Action.displayName = 'ActionList.Action';
Action.propTypes = {
  content: PropTypes.node,
  children: PropTypes.node,
  disabled: PropTypes.bool,
  /**
   * Same as hover styles.
   * Can be used for wrappers that manage focus within the menu, eg downshift
   */
  highlighted: PropTypes.bool,
  is: PropTypes.oneOf(['link', 'button']),
  selected: deprecate(PropTypes.bool, 'Use the checkbox component instead'),
  helpText: PropTypes.string,
};

export default Action;
