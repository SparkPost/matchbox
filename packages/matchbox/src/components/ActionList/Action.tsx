import React from 'react';
import { CheckBox } from '@sparkpost/matchbox-icons';
import { Box } from '../Box';
import { HelpText } from '../HelpText';
import { StyledLink } from './styles';
import { LinkActionProps } from '../../types/actions';

type BaseProps = {
  children?: React.ReactNode;
  disabled?: boolean;
  helpText?: React.ReactNode;
  highlighted?: boolean;

  /**
   * @deprecated Use Checkboxes instead
   */
  selected?: boolean;

  /**
   * @deprecated Use children instead
   */
  content?: React.ReactNode;

  /**
   * @deprecated Compose actions with ActionList.Section instead
   */
  group?: number | string;

  /**
   * @deprecated Compose actions with ActionList.Action instead
   */
  visible?: boolean;
};

type LinkProps = {
  is?: 'link';
} & LinkActionProps &
  React.ComponentPropsWithoutRef<'a'> &
  BaseProps;

type ButtonProps = {
  is: 'button';
} & React.ComponentPropsWithoutRef<'button'> &
  BaseProps;

type ActionProps = LinkProps | ButtonProps;

const Action = React.forwardRef<HTMLAnchorElement, ActionProps>(function Action(
  props: ActionProps,
  userRef,
) {
  const { content, children, disabled, helpText, is = 'link', selected, ...action } = props;

  const linkContent = React.useMemo(() => {
    return (
      <Box as="span" alignItems="flex-start" display="flex">
        <Box as="span" flex="1" fontSize="300" lineHeight="300">
          {content}
          {children}
        </Box>
        {selected && (
          <Box as="span" color="blue.700">
            <CheckBox size={20} />
          </Box>
        )}
        {helpText && <HelpText>{helpText}</HelpText>}
      </Box>
    );
  }, [content, selected, helpText]);

  return (
    <StyledLink
      as={is === 'button' ? 'button' : null}
      type={is === 'button' ? 'button' : null}
      disabled={disabled}
      isType={is}
      ref={userRef}
      role="menuitem"
      tabIndex="-1"
      {...action}
    >
      {linkContent}
    </StyledLink>
  );
});

Action.displayName = 'ActionList.Action';

export default Action;
