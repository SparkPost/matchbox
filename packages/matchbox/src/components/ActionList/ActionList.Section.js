import React, { Component } from 'react';
import classnames from 'classnames';
import { Check } from '@sparkpost/matchbox-icons';
import { linkFrom } from '../UnstyledLink';
import styles from './ActionList.module.scss';

const Section = ({ section }) => {
  const actions = section.map(({ className, highlighted, selected, content, ...action }, index) => {

    const classes = classnames(
      styles.Action,
      highlighted && styles.highlighted,
      className
    );

    const linkContent = selected
      ? <span>{content}<Check className={styles.Check} size={21}/></span>
      : content;

    return linkFrom({ content: linkContent, ...action, className: classes }, index);
  });

  return (
    <div className={styles.Section}>
      {actions}
    </div>
  );
};
Section.displayName = 'ActionList.Section';

export default Section;
