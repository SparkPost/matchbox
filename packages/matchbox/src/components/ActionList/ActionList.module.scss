@import '../../styles/config';

.ActionList {
  overflow-y: auto;
  z-index: 1;
}

.Section {
  border-bottom: 1px solid color(gray, 8);

  &:last-child {
    border: none;
  }

  &:first-child {
    margin-top: 3px;
  }
}

.Action, a.Action {
  display: block;
  padding: rem(9) spacing();
  color: color(gray, 4);
  transition: 0.15s;

  &:hover {
    background: color(gray, 9);
    color: color(gray, 1);
  }
}

.highlighted, a.highlighted {
  background: color(gray, 9);
  color: color(gray, 1);
}

.Check {
  float: right;
  margin-left: spacing();
  color: color(orange);
}
