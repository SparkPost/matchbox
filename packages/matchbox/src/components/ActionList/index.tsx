export { default as ActionList } from './ActionList';
export * from './ActionList';
export * from './Action';
export * from './Section';
