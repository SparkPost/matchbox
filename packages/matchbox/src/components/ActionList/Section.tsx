import React from 'react';
import Action from './Action';
import { filterByVisible } from '../../helpers/array';

import { StyledSection } from './styles';

export type SectionProps = {
  /**
   * @deprecated
   */
  section?: React.ComponentProps<typeof Action>[];
  children?: React.ReactNode;
};

const Section = React.forwardRef<HTMLDivElement, SectionProps>(function Section(
  props: SectionProps,
  userRef,
) {
  const { section, children } = props;

  if (children) {
    return <StyledSection ref={userRef}>{children}</StyledSection>;
  }

  const visibleActions = section
    ? filterByVisible(section).map((action, i) => <Action key={i} {...action} />)
    : [];

  if (!visibleActions.length) {
    return null;
  }

  return <StyledSection ref={userRef}>{visibleActions}</StyledSection>;
});

Section.displayName = 'ActionList.Section';

export default Section;
