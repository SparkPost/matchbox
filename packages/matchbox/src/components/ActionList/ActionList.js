import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { createPropTypes } from '@styled-system/prop-types';
import { margin, layout, compose } from 'styled-system';
import { groupByValues } from '../../helpers/array';
import { deprecate } from '../../helpers/propTypes';
import Section from './Section';
import Action from './Action';

const system = compose(margin, layout);
const Wrapper = styled('div')`
  ${system}
  overflow-y: auto;
`;

const ActionList = React.forwardRef(function ActionList(props, userRef) {
  const {
    actions = [],
    className,
    children,
    'data-id': dataId,
    sections = [],
    maxHeight = 'none',
    onClick,
    groupByKey = 'section',
    ...rest
  } = props;

  let list = actions && actions.length ? groupByValues(actions, groupByKey) : [];
  if (sections && sections.length) {
    list = list.concat(sections);
  }

  const listMarkup = list.map((section, index) => <Section section={section} key={index} />);

  return (
    <Wrapper
      className={className}
      data-id={dataId}
      maxHeight={typeof maxHeight === 'number' ? `${maxHeight}px` : maxHeight}
      onClick={onClick}
      ref={userRef}
      tabIndex="-1"
      {...rest}
    >
      {listMarkup}
      {children}
    </Wrapper>
  );
});

ActionList.displayName = 'ActionList';
ActionList.propTypes = {
  /**
   * Actions
   * e.g. [{ content: 'action label', onClick: callback() }]
   *
   * Note: each item can include an optional "section" key that will be used to auto group into sections, declaratively
   */
  actions: deprecate(
    PropTypes.arrayOf(PropTypes.shape({ content: PropTypes.node.isRequired })),
    'Use the ActionList.Action component instead',
  ),
  className: PropTypes.string,
  'data-id': PropTypes.string,
  /**
   * Creates sections
   * e.g. [[{ content: 'action label', onClick: callback() }]]
   */
  sections: deprecate(
    PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.shape({ content: PropTypes.node.isRequired }))),
    'Use the ActionList.Section component instead',
  ),

  /**
   * Max height of list
   */
  maxHeight: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onClick: PropTypes.func,

  /**
   * Group by key used to auto group actions into sections, defaults to "section"
   */
  groupByKey: deprecate(PropTypes.string, 'Use the ActionList.Section component instead'),
  children: PropTypes.node,
  ...createPropTypes(margin.propNames),
  ...createPropTypes(layout.propNames),
};

ActionList.Section = Section;
ActionList.Action = Action;
export default ActionList;
