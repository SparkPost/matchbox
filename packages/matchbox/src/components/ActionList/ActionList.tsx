import React from 'react';
import styled from 'styled-components';
import {
  margin,
  layout,
  compose,
  LayoutProps,
  MarginProps,
  maxHeight,
  MaxHeightProps,
} from 'styled-system';
import { groupByValues } from '../../helpers/array';
import Section from './Section';
import Action from './Action';
import { onKey } from '../../helpers/keyEvents';

export interface ActionListProps
  extends React.ComponentPropsWithRef<'div'>,
    LayoutProps,
    MarginProps,
    MaxHeightProps {
  children?: React.ReactNode;

  /**
   * @deprecated Use the ActionList.Action component instead
   */
  actions?: React.ComponentProps<typeof Action>[];

  /**
   * @deprecated Use the ActionList.Section component instead
   */
  sections?: React.ComponentProps<typeof Action>[][];

  /**
   * @deprecated Use the ActionList.Section component instead
   */
  groupByKey?: string;
  'aria-labelledby'?: string;
  'data-id'?: string;
}

const system = compose(margin, maxHeight, layout);

const Wrapper = styled.div<MaxHeightProps>`
  ${system}
  overflow-y: auto;
`;

const ActionList = React.forwardRef<HTMLDivElement, ActionListProps>(function ActionList(
  props,
  userRef,
) {
  const {
    actions = [],
    'aria-labelledby': labelledBy,
    className,
    children,
    'data-id': dataId,
    id,
    sections = [],
    maxHeight = 'none',
    onClick,
    groupByKey = 'section',
    ...rest
  } = props;

  // TODO Remove `actions` and `sections` support in favor of composable components
  let list = actions && actions.length ? groupByValues(actions, groupByKey) : [];
  if (sections && sections.length) {
    list = list.concat(sections);
  }

  const listMarkup = list.map((section, index) => <Section section={section} key={index} />);

  const wrapperRef = React.useRef(null);
  const [focusableItemList, setFocusableItemList] = React.useState([]);

  const [focusIndex, setFocusIndex] = React.useState(0);

  // Creates a list of focusable links or buttons inside the actionlist
  React.useEffect(() => {
    if (!!wrapperRef && wrapperRef.current) {
      setFocusableItemList(
        wrapperRef.current.querySelectorAll(
          '[role="menuitem"]:not(:disabled):not([aria-disabled="true"])',
        ),
      );
    }
  }, []);

  React.useLayoutEffect(() => {
    if (focusableItemList[focusIndex]) {
      // Honestly not sure why this doesn't work without a timeout
      setTimeout(() => {
        focusableItemList[focusIndex].focus();
      }, 10);
    }
  }, [focusIndex, focusableItemList]);

  function handleKeyDown(e: React.KeyboardEvent) {
    onKey('arrowDown', () => {
      // Stop arrow keys from scrolling the page
      e.preventDefault();

      if (focusIndex < focusableItemList.length - 1) {
        setFocusIndex(focusIndex + 1);
      } else {
        setFocusIndex(0);
      }
    })(e);

    onKey('arrowUp', () => {
      // Stop arrow keys from scrolling the page
      e.preventDefault();

      if (focusIndex <= 0) {
        setFocusIndex(focusableItemList.length - 1);
      } else {
        setFocusIndex(focusIndex - 1);
      }
    })(e);
  }

  function assignRefs(node) {
    wrapperRef.current = node;
    if (userRef) {
      (userRef as React.MutableRefObject<HTMLDivElement>).current = node;
    }
  }

  return (
    <Wrapper
      aria-labelledby={labelledBy}
      className={className}
      data-id={dataId}
      id={id}
      maxHeight={typeof maxHeight === 'number' ? `${maxHeight}px` : maxHeight}
      onClick={onClick}
      tabIndex={-1}
      role="menu"
      ref={assignRefs}
      onKeyDown={handleKeyDown}
      {...rest}
    >
      {listMarkup}
      {children}
    </Wrapper>
  );
}) as React.ForwardRefExoticComponent<ActionListProps> & {
  Section: typeof Section;
  Action: typeof Action;
};

ActionList.displayName = 'ActionList';

ActionList.Section = Section;
ActionList.Action = Action;
export default ActionList;
