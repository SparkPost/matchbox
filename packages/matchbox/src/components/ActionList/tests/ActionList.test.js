import React from 'react';
import ActionList from '../ActionList';
import { shallow } from 'enzyme';

describe('ActionList', () => {
  it('renders actions correctly', () => {
    const wrapper = shallow(
      <ActionList actions={[{ content: 'Action' }, { content: 'Action 2', selected: true }]} />
    );
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('Section').dive()).toMatchSnapshot();
  });

  it('renders sections correctly', () => {
    const wrapper = shallow(
      <ActionList
        sections={[
          [{ content: 'Sectioned1' }, { content: 'Sectioned2' }],
          [{ content: 'Sectioned3' }, { content: 'Sectioned4' }]
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders sections correctly using a group by key', () => {
    const wrapper = shallow(
      <ActionList
        actions={[
          { content: 'Sectioned1', section: 1 }, { content: 'Sectioned2', section: 1 },
          { content: 'Sectioned3', section: 2 }, { content: 'Sectioned4', section: 2 }
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders sections with actions and max height correctly', () => {
    const wrapper = shallow(
      <ActionList
        maxHeight={50}
        actions={[{ content: 'Action1' }, { content: 'Action2' }]}
        sections={[
          [{ content: 'Sectioned1' }, { content: 'Sectioned2' }],
          [{ content: 'Sectioned3' }, { content: 'Sectioned4' }]
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
