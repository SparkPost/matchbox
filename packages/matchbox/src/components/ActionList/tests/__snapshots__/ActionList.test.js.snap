// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionList renders actions correctly 1`] = `
<div
  className="ActionList"
  style={
    Object {
      "maxHeight": "none",
    }
  }
>
  <Section
    key="0"
    section={
      Array [
        Object {
          "content": "Action",
        },
        Object {
          "content": "Action 2",
          "selected": true,
        },
      ]
    }
  />
</div>
`;

exports[`ActionList renders actions correctly 2`] = `
<div
  className="Section"
>
  <UnstyledLink
    className="Action"
    key="0"
  >
    Action
  </UnstyledLink>
  <UnstyledLink
    className="Action"
    key="1"
  >
    <span>
      Action 2
      <Check
        className="Check"
        size={21}
      />
    </span>
  </UnstyledLink>
</div>
`;

exports[`ActionList renders sections correctly 1`] = `
<div
  className="ActionList"
  style={
    Object {
      "maxHeight": "none",
    }
  }
>
  <Section
    key="0"
    section={
      Array [
        Object {
          "content": "Sectioned1",
        },
        Object {
          "content": "Sectioned2",
        },
      ]
    }
  />
  <Section
    key="1"
    section={
      Array [
        Object {
          "content": "Sectioned3",
        },
        Object {
          "content": "Sectioned4",
        },
      ]
    }
  />
</div>
`;

exports[`ActionList renders sections correctly using a group by key 1`] = `
<div
  className="ActionList"
  style={
    Object {
      "maxHeight": "none",
    }
  }
>
  <Section
    key="0"
    section={
      Array [
        Object {
          "content": "Sectioned1",
          "section": 1,
        },
        Object {
          "content": "Sectioned2",
          "section": 1,
        },
      ]
    }
  />
  <Section
    key="1"
    section={
      Array [
        Object {
          "content": "Sectioned3",
          "section": 2,
        },
        Object {
          "content": "Sectioned4",
          "section": 2,
        },
      ]
    }
  />
</div>
`;

exports[`ActionList renders sections with actions and max height correctly 1`] = `
<div
  className="ActionList"
  style={
    Object {
      "maxHeight": 50,
    }
  }
>
  <Section
    key="0"
    section={
      Array [
        Object {
          "content": "Action1",
        },
        Object {
          "content": "Action2",
        },
      ]
    }
  />
  <Section
    key="1"
    section={
      Array [
        Object {
          "content": "Sectioned1",
        },
        Object {
          "content": "Sectioned2",
        },
      ]
    }
  />
  <Section
    key="2"
    section={
      Array [
        Object {
          "content": "Sectioned3",
        },
        Object {
          "content": "Sectioned4",
        },
      ]
    }
  />
</div>
`;
